File: CPP/RaceCar.cpp
#include <iostream>
#include <queue>
#include <set>
#include <tuple>

class Solution {
public:
  int racecar(int target) {
    int bound = 2 * target;
    std::set<std::pair<int, int>> seen{{0, 1}};
    std::queue<std::tuple<int, int, int>> q; // pos, speed, steps
    q.push({0, 1, 0});
    while (!q.empty()) {
      auto [pos, speed, steps] = q.front();
      q.pop();
      if (pos == target)
        return steps;
      // 'A'
      int npos = pos + speed, nspeed = speed * 2;
      if (npos >= -bound && npos <= bound &&
          seen.insert({npos, nspeed}).second) {
        q.push({npos, nspeed, steps + 1});
      }
      // 'R'
      int rspeed = speed > 0 ? -1 : 1;
      if (seen.insert({pos, rspeed}).second) {
        q.push({pos, rspeed, steps + 1});
      }
    }
    return -1;
  }
};

int main() {
  Solution s;
  std::cout << s.racecar(3) << "\n"; // 2
  std::cout << s.racecar(6) << "\n"; // 5
  std::cout << s.racecar(1) << "\n"; // 1
  std::cout << s.racecar(4) << "\n"; // 5
  return 0;
}

File: Rust/src/bin/race_car.rs
use std::collections::{HashSet, VecDeque};

fn racecar(target: i32) -> i32 {
    let bound = 2 * target;
    let mut seen: HashSet<(i32, i32)> = HashSet::new();
    seen.insert((0, 1));
    let mut queue: VecDeque<(i32, i32, i32)> = VecDeque::new();
    queue.push_back((0, 1, 0));
    while let Some((pos, speed, steps)) = queue.pop_front() {
        if pos == target {
            return steps;
        }
        // 'A'
        let (npos, nspeed) = (pos + speed, speed * 2);
        if npos >= -bound && npos <= bound && seen.insert((npos, nspeed)) {
            queue.push_back((npos, nspeed, steps + 1));
        }
        // 'R'
        let rspeed = if speed > 0 { -1 } else { 1 };
        if seen.insert((pos, rspeed)) {
            queue.push_back((pos, rspeed, steps + 1));
        }
    }
    -1
}

fn main() {
    println!("{}", racecar(3)); // 2
    println!("{}", racecar(6)); // 5
    println!("{}", racecar(1)); // 1
    println!("{}", racecar(4)); // 5
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert_eq!(racecar(3), 2);
        assert_eq!(racecar(6), 5);
        assert_eq!(racecar(1), 1);
        assert_eq!(racecar(4), 5);
    }
}
